Add c10d filestore functionality to the current c10d_rendezvous_backend

SirLegolot opened the pull request:
Summary: Added filestore functionality to the c10d backend. FileStore will create a temporary file in the /tmp directory to use if it is selected as the store type. Appropriate tests were added as well.

Test Plan: Unit tests for creating a c10d backend with filestore and simple error handling.

Differential Revision: D28997436

cbalioglu commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -198,9 +216,12 @@ def create_backend(params: RendezvousParameters) -> Tuple[C10dRendezvousBackend,
-        raise ValueError("The store type must be 'tcp'. Other store types are not supported yet.")
 
-    store = _create_tcp_store(params)
+    if store_type == "filestore":
```
I would name this just "file" to be consistent with "tcp".

H-Huang commented on test/distributed/elastic/rendezvous/c10d_rendezvous_backend_test.py:
```diff
@@ -36,9 +39,34 @@ def setUp(self) -> None:
+        self._store.set("torch.rendezvous.dummy_run_id", "non_base64")
+
 
 class CreateBackendTest(TestCase):
     def setUp(self) -> None:
+        # These are the default parameters, which are tcp
```
Since the default is to run `CreateBackendTest` with TCP, instead of adding separate test cases only for FileStore in this class and having multiple conditionals to separate the two, can we instead use inheritance? E.g.

```python
class CreateBackendTest(TestCase):
    # as before...

class CreateBackendTestFileStore(CreateBackendTest):
   def setUp(self):
       # override setUp method by setting self.params to be your filestore params

   def test_bad_min_and_max_nodes_filestore(self):
       # your implementation (I believe this was the only unique test for filestore that was added)
```
This would run the same common suite of tests for TCP and FileStore, as well as the additionally added one for FileStore.

Alternatively, you could split it up into a common class such as `CreateBackendTestCommon` and have `CreateBackendTestTCP` and `CreateBackendTestFileStore` both inherit from it, but that may be overkill.

cbalioglu commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -166,6 +167,23 @@ def _create_tcp_store(params: RendezvousParameters) -> TCPStore:
+    # up after use. In the future, a close method for rendezvous backend should
+    # be implemented to enable cleaning up resources. (For now, since it
+    # writes to /tmp, it is ok)
+    temp_dir = tempfile.mkdtemp(prefix="torch_rdzv_")
```
I would simplify these two lines just to a `_, path = tempfile.mkstemp()`. This is an implementation specific temporary binary file that gets deleted at the end of the job. No need to have a directory structure or a naming convention.

cbalioglu replied:
Make sure to wrap any exceptions raised from these file system calls in a `RendezvousError`.

```python
try:
    _, path = tempfile.mkstemp()
except OSError as exc:
   raise RendezvousError("failed!") from exc
```

cbalioglu commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -166,6 +167,23 @@ def _create_tcp_store(params: RendezvousParameters) -> TCPStore:
+
+    # Min and max number of nodes should be the same so we can ensure
+    # num_workers, which is passed in to create the FileStore is correct.
+    if (params.min_nodes != params.max_nodes):
```
This defeats the whole purpose of dynamic rendezvous handling. I encourage you to lift this limitation from `FileStore`. I just checked its source code and apparently the reason why we require `numWorkers` is to delete the file once all workers are complete (see [here](https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/FileStore.cpp#L286)).

Since we are already using a temporary file, there is really no need for us to explicitly delete it. Ideally we should make the `numWorkers` parameter of `FileStore` optional and skip the file deletion in the constructor if it is not set.

cbalioglu commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -198,9 +216,12 @@ def create_backend(params: RendezvousParameters) -> Tuple[C10dRendezvousBackend,
     # As of today we only support TCPStore. Other store types do not have the
```
This comment needs to be updated.

cbalioglu reviewed (commented):
So far I have only reviewed the actual implementation. I have a few comments and a request for an improvement in `FileStore`. Once we settle down the implementation part, I will review the tests.

SirLegolot commented on test/distributed/elastic/rendezvous/c10d_rendezvous_backend_test.py:
```diff
@@ -50,43 +78,76 @@ def setUp(self) -> None:
+        if (self._expected_store_type == TCPStore):
+            self.assertEqual(tcp_store.host, self._expected_endpoint_host)  # type: ignore[attr-defined]
+            self.assertEqual(tcp_store.port, self._expected_endpoint_port)  # type: ignore[attr-defined]
+            self.assertEqual(tcp_store.timeout, self._expected_read_timeout)  # type: ignore[attr-defined]
```
So for that, FileStore does not take in timeout as an option (which makes sense, since the store is based on a local file). In testing, it seems that the Store timeout property ends up defaulting to 300 seconds instead of the 60 seconds that it says it should be in documentation. For that reason I had left it out of FileStore tests.

H-Huang replied:
Yes, `FileStore` does not have a timeout argument in its constructor, but all classes derived from `Store` have the timeout property. The timeout is still used in FileStore for all operations when accessing the file, e.g. (`get`, `est`, `add`, etc.), here's a pointer if curious https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/FileStore.cpp#L154-L158. Also wondering where in the docs it says 60 seconds, probably an error, can you link to me?

SirLegolot replied:
Ah ok. Here's the documentation I was referring to: https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py#L182
I believe that it said 60 there because there was some logic in _create_tcp_store that defaults it to 60 if not specified:
https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py#L137

H-Huang replied:
I see, thought you meant official pytorch docs. You can also set timeout to be 60 in your internal `_create_file_store` so it matches.

cbalioglu reviewed (commented):
I have one relatively important request. Most of my comments are minor.

SirLegolot commented on test/distributed/elastic/rendezvous/c10d_rendezvous_backend_test.py:
```diff
@@ -36,9 +42,34 @@ def setUp(self) -> None:
+
+    @classmethod
+    def setUpClass(cls):
+        temp_dir = tempfile.mkdtemp(prefix="torch_rdzv_")
```
I was thinking that if we did it this way, we can clean up the temporary files that are created, especially if we change the initialization of FileStore to drop the numWorkers_ parameter (e.g. FileStore(path, 1) -> FileStore(path)), in which case the file is not cleaned up in the end.

SirLegolot commented on torch/lib/c10d/FileStore.cpp:
```diff
@@ -272,18 +272,16 @@ FileStore::FileStore(const std::string& path, int numWorkers)
+  // The last worker cleans up the file. If numWorkers was not initialized to
+  // a specific postive value (i.e. meaning that there was not a fixed number
+  // of workers), we don't attempt to clean.
+  if (numWorkers_ > 0 && numFinishedWorker == numWorkers_) {
```
i guess not. A user could technically instantiate a FileStore object with 0, do nothing with it, and then it gets deleted without deleting the file created.

H-Huang commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -807,7 +807,7 @@ A store implementation that uses a file to store the underlying key-value pairs.
 Arguments:
     file_name (str): path of the file in which to store the key-value pairs
-    world_size (int): The total number of processes using the store
+    world_size (int, optional): The total number of processes using the store. Default is -1 (a negative value indicates an non-fixed number of store users).
```
typo: "an" - > "a"

H-Huang reviewed (approved):
Changes look good, thanks for updating!

cbalioglu commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -166,6 +167,21 @@ def _create_tcp_store(params: RendezvousParameters) -> TCPStore:
+                "The file creation for C10d store has failed. See inner exception for details."
+            ) from exc
+
+    return FileStore(path)
```
Since the endpoint can now be specified by the user as well, we should wrap this call with an exception handler similar to line 179.

SirLegolot replied:
So right now, it doesn't seem that there is any check on the path that is performed when FileStore is initialized (from looking at FileStore.cpp). It would only fail if an operation such as get/set is attempted to be performed. I'll still wrap this with a runtime/value error, but I wasn't sure.

cbalioglu commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -4,16 +4,17 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
+import tempfile
```
Nit: The import statements are not alphabetically sorted. We use [black](https://github.com/psf/black) for consistent formatting of our Python files. Make sure you run it before your PRs:

```
$ black --line-length 100 <path_to_file>
```

cbalioglu commented on test/distributed/elastic/rendezvous/c10d_rendezvous_backend_test.py:
```diff
@@ -97,26 +168,56 @@ def test_create_backend_returns_backend_if_is_host_is_not_specified_and_store_al
+                )
+
+    def test_create_backend_raises_error_if_store_type_not_specified_with_file_endpoint(self) -> None:
+        with self.assertRaises(ValueError):
```
Please see my comment below. We should not expose a `ValueError` in our API.

cbalioglu commented on test/distributed/elastic/rendezvous/c10d_rendezvous_backend_test.py:
```diff
@@ -50,43 +78,86 @@ def setUp(self) -> None:
+            self._expected_read_timeout = timedelta(seconds=300)
+            # in case there is a test which defaults to tcp store, we set the
+            # default tcp parameters
+            self._expected_endpoint_port = 29400
```
Nit: Is there such a test case? I would assume that if the store type is `file` and somehow we use a TCP port, it is likely a bug.

cbalioglu commented on test/distributed/elastic/rendezvous/c10d_rendezvous_backend_test.py:
```diff
@@ -97,26 +168,56 @@ def test_create_backend_returns_backend_if_is_host_is_not_specified_and_store_al
+        # test_create_backend_raises_error_if_store_type_not_specified_with_file_endpoint
+        self._params_filestore.endpoint = ""
+
+        for store_type in ["tcp", "file"]:
```
I think this test is superfluous. In the original test (lines from 185 to 192) we explicitly delete the `store_type` config before assertion. Not sure if we gain any test coverage by testing it with both store types here.

cbalioglu reviewed (commented):
Mostly looks good. I just have a few minor comments. We are almost there :)

Also please make sure that you run `flake8` and `black` to lint and format your code.

SirLegolot committed:
Add c10d filestore functionality to the current c10d_rendezvous_backend (#59719)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/59719

Added filestore functionality to the c10d backend. FileStore will create a temporary file in the /tmp directory to use if it is selected as the store type. Appropriate tests were added as well.
FileStore was modified to expose the path field for testing. It was also modified so that the numWorkers field in the constructor is optional (defaulting to -1). A negative value indicates there is not a fixed number of workers. In this case, the file is not attempted to be cleaned at the end.

Test Plan: Unit tests for creating a c10d backend with filestore and simple error handling.

Reviewed By: H-Huang

Differential Revision: D28997436

fbshipit-source-id: 9a13c158e5b28b145b3df7dd8ecbca688eda0ff8

cbalioglu reviewed (approved):
LGTM! Congrats on your first major contribution!

DTorres36 commented:
pog
